File: C/readability/readability.c
// Write a program called readability that takes a text and determines its reading level.
// Problem link: https://cs50.harvard.edu/x/2022/psets/2/readability/

#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

int count_letters(string text);
int count_words(string text);
int count_sentences(string text);

int main(void)
{
    // Prompt user for text
    string text = get_string("Text: ");

    // Number of letters, words and sentences in text
    int letters = count_letters(text);
    int words = count_words(text);
    int sentences = count_sentences(text);

    // Average number of letters per 100 words in the text
    float avgLetters = (letters / (float)words) * 100;
    // Average number of sentences per 100 words in the text
    float avgSentences = (sentences / (float)words) * 100;

    // Compute Coleman-Liau index (it outputs the U.S. grade level)
    int index = round(0.0588 * avgLetters - 0.296 * avgSentences - 15.8);

    // Print out grade level
    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", index);
    }
}

// Calculate number of letters in text
int count_letters(string text)
{
    int letters = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }
    return letters;
}

// Calculate number of words in text
int count_words(string text)
{
    int spaces = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == ' ')
        {
            spaces++;
        }
    }
    int words = spaces + 1;
    return words;
}

// Calculate number of sentences in text
int count_sentences(string text)
{
    int sentences = 0;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
    }
    return sentences;
}

File: python/sentimental-readability/readability.py
# Implement a program that computes the approximate grade level needed to comprehend some text
# Problem link: https://cs50.harvard.edu/x/2022/psets/6/readability/

from cs50 import get_string
import re

# Get the user’s input
text = get_string("Text: ")


# Keep count of letters, words and sentences
# (add 1 to words for the last word)
n_letters = 0
n_words = 1
n_sentences = 0


for letter in text:
    # Count letters
    if re.match("[a-z]|[A-Z]", letter):
        n_letters += 1
    # Count white spaces to count words
    if letter == " ":
        n_words += 1
    # Count sentences (consider any sequence of characters that ends with a . or a ! or a ? to be a sentence)
    if letter == "." or letter == "!" or letter == "?":
        n_sentences += 1


# Apply the Coleman-Liau index
# index = 0.0588 * L - 0.296 * S - 15.8 (L = average number of letters per 100 words, S = average number of sentences per 100 words)
avg_letters = (n_letters / n_words) * 100
avg_sentences = (n_sentences / n_words) * 100

idx = round(0.0588 * avg_letters - 0.296 * avg_sentences - 15.8)

# Print result
if idx < 1:
    print("Before Grade 1")
elif idx >= 16:
    print("Grade 16+")
else:
    print(f"Grade {idx}")
